### PR TITLE
Implement List or map: remember last used #649

### PR DESCRIPTION
**Pull request by mashawan**
I introduced a MAP/LIST enum, and used this to determine which mode we are in, then toggle if the map/list button is pressed. When Switching modes a shared preference is updated, this is used to initial set the current mode when opening the activity.

### CODE REVIEW

File: app/src/main/res/values-hi/strings.xml [hunk 1]
```diff
@@ -54,7 +54,10 @@
   <string name="gps_disabled">आपके डिवाइस में जीपीएस अक्षम है। क्या आप इसे सक्षम करना चाहेंगे?</string>
   <string name="enable_gps">जीपीएस सक्षम करें</string>
   <string name="contributions_subtitle_zero">अभी तक कोई अपलोड नहीं</string>
-  <string name="contributions_subtitle">{{PLURAL | शून्य = @ स्ट्रिंग / योगदान_उपशीर्षक_शुन्य | एक =% d अपलोड |% d अपलोड्स}}</string>
+  <plurals name="contributions_subtitle">
+    <item quantity="one">%d अपलोड</item>
+    <item quantity="other">%d अपलोड्स</item>
+  </plurals>
```
Comment (nicolas-raoul) [inline app/src/main/res/values-hi/strings.xml]: Is this related?

**Comment by nicolas-raoul** [comment]
I haven't tested, but it looks good :-)

File: app/src/main/res/values-hi/strings.xml (hunk 1 above)
Comment (mashawan) [reply app/src/main/res/values-hi/strings.xml]: I'm not sure what happened here, I rebased to latest and got failing builds, so I needed to fix this in order to continue. Looks like a translation update added a spurious string entry...

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 2]
```diff
@@ -42,28 +44,46 @@
 
     @BindView(R.id.progressBar)
     ProgressBar progressBar;
-    private boolean isMapViewActive = false;
     private static final int LOCATION_REQUEST = 1;
+    public static final String MAP_LAST_USED_PREFERENCE = "mapLastUsed";
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: This is only used in the context of NearbyActivity and could be private.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 3]
```diff
@@ -42,28 +44,46 @@
...
+        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
         setContentView(R.layout.activity_nearby);
         ButterKnife.bind(this);
         checkLocationPermission();
         bundle = new Bundle();
         initDrawer();
+        initViewMode();
+    }
+
+    private void initViewMode() {
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: If you were to store this preference as a String, you could use the built-in "valueOf()" method of enum.  Alternatively, this method should probably move to the NearbyActivityMode class.  In either case initializing the viewMode could be one line of code.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 4]
```diff
@@ -42,28 +44,46 @@
...
+            viewMode = NearbyActivityMode.LIST;
+        }
     }
 
     @Override
     public boolean onCreateOptionsMenu(Menu menu) {
         MenuInflater inflater = getMenuInflater();
         inflater.inflate(R.menu.menu_nearby, menu);
+
+        if (viewMode == NearbyActivityMode.MAP) {
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: You do this test several times, maybe add an "isMap()" on the enum?

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 5]
```diff
@@ -74,19 +94,23 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+            case R.id.action_toggle_view:
+                toggleView();
+                toggleIcon(item);
                 return true;
             default:
                 return super.onOptionsItemSelected(item);
         }
     }
 
+    private void toggleIcon(MenuItem item) {
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: This can be one line of code - if you use a ternary expression, or if you added a 'getIcon()' to the enum and initialized the two values with resource ID integers.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 6]
```diff
@@ -198,23 +222,29 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         }
     }
 
-    private void showMapView() {
+    private void toggleView() {
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: If you added a 'toggle()' method on the enum, then the complexity of this method actually falls away
```java
if (nearbyAsyncTask != null) {
    viewMode = viewMode.toggle();
    if (nearbyAsyncTask.getStatus() == AsyncTask.Status.FINISHED) {
        if (viewMode.isMap()) {
            setMapFragment();
        } else {
            setListFragment();
        }
    }
    sharedPreferences.edit().putBoolean(MAP_LAST_USED_PREFERENCE, viewMode.isMap()).apply();
}
```

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java [hunk 7]
```diff
@@ -198,23 +222,29 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                viewMode = NearbyActivityMode.LIST;
                 if (nearbyAsyncTask.getStatus() == AsyncTask.Status.FINISHED) {
                     setListFragment();
+                    setMapViewLastUsed(false);
                 }
             }
         }
     }
 
+    private void setMapViewLastUsed(boolean lastUsed) {
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivity.java]: Possibly not needed & could be deleted, depending on other changes.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivityMode.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package fr.free.nrw.commons.nearby;
+
+enum NearbyActivityMode {
+    MAP,
+    LIST
+}
```
Comment (psh) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyActivityMode.java]: If you run with the suggestions from NearbyActivity, this class would look quite different -
```java
enum NearbyActivityMode {
    MAP(R.drawable.ic_list_white_24dp),
    LIST(R.drawable.ic_map_white_24dp);

    @DrawableRes
    private final int icon;

    NearbyActivityMode(int icon) {
        this.icon = icon;
    }

    public NearbyActivityMode toggle() {
        return isMap() ? LIST : MAP;
    }

    public boolean isMap() {
        return MAP.equals(this);
    }

    public static NearbyActivityMode fromBoolean(boolean value) {
        return value ? MAP : LIST;
    }

    @DrawableRes
    public int getIcon() {
        return icon;
    }
}
```

**Comment by mashawan** [comment]
Thanks for the feed back, I will consider these suggestions and see if I can improve on what I have so far.  I'll try and push to my fork in the next few days.

**Comment by mashawan** [comment]
Code updated with suggestions.  I think this is done now.

**Comment by psh** [comment]
Looks good to me, thanks for the contribution!
